internal/codycontext/dotcom.go

```
@@ -0,0 +1,167 @@
+package codycontext
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/zoekt/ignore"
+
+	"github.com/sourcegraph/sourcegraph/internal/actor"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/gitserver"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+)
+
+const codyIgnoreFile = ".cody/ignore"
+
+var (
+	emptyMatcher ignore.Matcher = ignore.Matcher{}
+)
+
+type filterFunc func(string) bool
+
+type repoRevision struct {
+	Repo   types.RepoIDName
+	Commit api.CommitID
+}
+
+type dotcomRepoFilter struct {
+	cache  safeCache[repoRevision, ignore.Matcher]
+	client gitserver.Client
+
+	mu      sync.RWMutex
+	enabled bool
+}
+
+func (f *dotcomRepoFilter) SetEnabled(enabled bool) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.enabled = enabled
+}
+
+func (f *dotcomRepoFilter) GetEnabled() bool {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.enabled
+}
+
+func (f *dotcomRepoFilter) GetFilter(repos []types.RepoIDName, logger log.Logger) ([]types.RepoIDName, FileChunkFilterFunc) {
+	if !f.GetEnabled() {
+		return repos, func(fcc []FileChunkContext) []FileChunkContext {
+			return fcc
+		}
+	}
+	return f.getFilter(repos, logger)
+}
+
+// getFilter returns the list of repos that can be filtered
+// their .cody/ignore files (or don't have one). If an error
+// occurs that repo will be excluded.
+func (f *dotcomRepoFilter) getFilter(repos []types.RepoIDName, logger log.Logger) ([]types.RepoIDName, FileChunkFilterFunc) {
```

how big can this list of repos be? The code below won't scale to large sizes, so just checking this is relatively small.

---

I don't expect this list to be big. 
E.g., for VSCode we query for a list of automatically (repos from workspace) and manually added (repo selector) repos. Not sure about the exact number of repos user can add manually (I'd be very surprised if more than 10) and if PLG users even have those repo selector (this `getFilter` will be used only for dotcom, and I currently don't have it in my VSCode). 
We also limit the number of code and text results in the context search query (VSCode extension currently [queries for 15 code and 5 text](https://github.com/sourcegraph/cody/blob/a109df44a5fb4501fea923dbaa71fb794d188665/lib/shared/src/sourcegraph-api/graphql/client.ts#L514-L515) results).  
I think the JetBrains extension has similar logic.  

Though we don't limit the number of repos in the [resolver](https://github.com/sourcegraph/sourcegraph/blob/341998d8749e96caedc1744dc36faec9967f50a4/cmd/frontend/internal/context/resolvers/context.go#L31) so if a consumer queries the endpoint with big number of repos and code/text results count, the efficiency/scalability problem will become more obvious.

---

I think as a follow-up PR you should enforce a hard limit somewhere and document why. Indeed it does sound like the number should be relatively small so this is acceptable. But the hard limit will save your ass further down the line and pre-empt conversations around if we want larger contexts.

---

Do you mean that the [resolver](https://github.com/sourcegraph/sourcegraph/blob/341998d8749e96caedc1744dc36faec9967f50a4/cmd/frontend/internal/context/resolvers/context.go#L31) should return a descriptive error if the repos number passed in args is greater than N?

---

yes

---

https://github.com/sourcegraph/sourcegraph/issues/61878